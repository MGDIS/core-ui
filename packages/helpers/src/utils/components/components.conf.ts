import type { Page } from './components';

/**
 * Define getPage type
 */
export type IGetPage<T> = (offset?: number, filter?: Parameters<Array<T>['filter']>[0]) => Page<T>;

/**
 * Cursor type
 */
export type CursorType = 'first' | 'next' | 'previous' | 'last';
