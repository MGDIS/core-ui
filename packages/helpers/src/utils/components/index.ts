export * from './components';
export type { CursorType } from './components.conf';
