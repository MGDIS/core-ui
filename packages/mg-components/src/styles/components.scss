// prevent conflict between [hide] component global selector and mg-modal hide rendered attribute
[hide]:not(mg-modal) {
  display: none;
}
