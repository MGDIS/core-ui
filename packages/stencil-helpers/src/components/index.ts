export {
  createID,
  isValideID,
  formatID,
  ClassList,
  allItemsAreString,
  isTagName,
  focusableElements,
  getWindows,
  getParentWindows,
  isValidString,
  toString,
  isValidNumber,
  cleanString,
  nextTick,
  isObject,
  getObjectValueFromKey,
  type CursorType,
  Cursor,
  Page,
  Paginate,
} from '@mgdis/core-ui-helpers/dist/utils';
