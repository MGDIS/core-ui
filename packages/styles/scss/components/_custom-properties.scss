/**
  * Components custom properties
  * Must start with `--mg-c-`
*/
:root {
  /**
   * Atoms
   */

  // mg-badge
  --mg-c-badge-size: var(--mg-b-size-2w);
  --mg-c-badge-font-size: 1.1rem;

  // mg-button
  --mg-c-button-border-radius: var(--mg-b-size-radius-legacy);
  --mg-c-button-icon-border-radius: var(--mg-b-size-min-height);
  --mg-c-button-disabled-opacity: var(--mg-b-disabled-opacity);
  --mg-c-button-gradient: 1; /* Can only be 1 or 0 to enable disable gradient */
  --mg-c-button-border-variation: 1; /* Can only be 1 or 0 to enable disable border variation */

  // mg-button Primary
  --mg-c-button-primary-color-h: var(--mg-b-color-dark-h);
  --mg-c-button-primary-color-s: var(--mg-b-color-dark-s);
  --mg-c-button-primary-color-l: var(--mg-b-color-dark-l);
  --mg-c-button-primary-font-color: var(--mg-b-color-neutral);

  // mg-button Secondary
  --mg-c-button-secondary-color-h: var(--mg-b-color-neutral-h);
  --mg-c-button-secondary-color-s: var(--mg-b-color-neutral-s);
  --mg-c-button-secondary-color-l: var(--mg-b-color-neutral-l);
  --mg-c-button-secondary-font-color: var(--mg-b-color-dark);

  // mg-card
  --mg-c-card-padding-default: var(--mg-b-size-2w);
  --mg-c-card-border-radius-default: var(--mg-b-size-radius);
  --mg-c-card-background-default: var(--mg-b-color-light);
  --mg-c-card-border-default: var(--mg-b-size-border) solid hsl(var(--mg-b-color-dark-h) var(--mg-b-color-dark-s) var(--mg-b-color-dark-l) / 5%);
  --mg-c-card-box-shadow-default: var(--mg-b-box-shadow);

  // mg-divider
  --mg-c-divider-background-color: hsl(var(--mg-b-color-dark-h) var(--mg-b-color-dark-s) var(--mg-b-color-dark-l) / 15%);
  --mg-c-divider-thickness: var(--mg-b-size-border);
  --mg-c-divider-vertical-spacing: var(--mg-b-size-5w);

  // mg-icon
  --mg-c-icon-small-size: var(--mg-b-size-3v);
  --mg-c-icon-regular-size: var(--mg-b-size-2w);
  --mg-c-icon-medium-size: var(--mg-b-size-5v);
  --mg-c-icon-large-size: var(--mg-b-size-3w);
  --mg-c-icon-extra-large-size: var(--mg-b-size-4w);
  --mg-c-icon-border-radius: 50%;

  // mg-tag
  --mg-c-tag-height: var(--mg-b-size-3w);
  --mg-c-tag-border-radius: var(--mg-b-size-radius);
  --mg-c-tag-font-size: 1.2rem;

  // mg-tooltip
  --mg-c-tooltip-border-radius: var(--mg-b-size-radius);
  --mg-c-tooltip-background-color: var(--mg-b-color-dark);
  --mg-c-tooltip-font-color: var(--mg-b-color-font-light);

  /**
   * Molecules
   */

  // mg-details
  --mg-c-details-spacing: var(--mg-b-size-w);

  // mg-illustrated-message
  --mg-c-illustrated-message-padding-vertical: var(--mg-b-size-5w);

  // mg-input-...
  --mg-c-input-border-width: var(--mg-b-size-border);
  --mg-c-input-border-radius: var(--mg-b-size-radius-legacy);
  --mg-c-input-color: hsl(201deg 16% 75%);
  --mg-c-input-spacer: var(--mg-b-size-w);

  /* Error message background */
  --mg-c-input-error-bg-color: hsl(var(--mg-b-color-danger-h) calc(var(--mg-b-color-danger-s) + 5%) calc(var(--mg-b-color-danger-l) + 49%));

  /* Input focus shadow */
  --mg-c-input-color-shadow-focus: hsl(188deg 100% 50% / 50%);

  // mg-input-toggle
  --mg-c-input-toggle-border-radius-ratio: 2;

  // mg-input-checkbox
  --mg-c-input-check-size: var(--mg-b-size-2w);

  // mg-menu-item
  --mg-c-menu-item-focused-background-color: hsl(var(--mg-b-color-app-h) var(--mg-b-color-app-s) var(--mg-b-color-app-l) / 10%);
  --mg-c-menu-item-border-color-active: var(--mg-b-color-app);
  --mg-c-menu-item-color-active: var(--mg-b-color-app);

  // mg-message
  --mg-c-message-border-radius: var(--mg-b-size-radius);

  // mg-modal
  --mg-c-modal-border-radius: var(--mg-b-size-radius);
  --mg-c-modal-title-font-size: var(--mg-b-font-size-h3);
  --mg-c-modal-content-font-size: 1.2rem;

  // mg-panel
  --mg-c-panel-border-radius: var(--mg-b-size-radius);
  --mg-c-panel-background: hsl(var(--mg-b-color-info-h) var(--mg-b-color-info-s) calc(var(--mg-b-color-info-l) + 68%));
  --mg-c-panel-box-shadow: var(--mg-b-box-shadow);
  --mg-c-panel-content-padding: var(--mg-b-size-2w);

  // mg-popover
  --mg-c-popover-background-color: var(--mg-b-color-light);
  --mg-c-popover-font-color: var(--mg-b-color-font-dark);
  --mg-c-popover-title-font-size: var(--mg-b-font-size-h5);
  --mg-c-popover-max-width: var(--mg-b-size-floating-element-max-width);

  // mg-tabs
  --mg-c-tabs-focused-background-color: hsl(var(--mg-b-color-app-h) var(--mg-b-color-app-s) var(--mg-b-color-app-l) / 10%);
  --mg-c-tabs-border-color-active: var(--mg-b-color-app);
  --mg-c-tabs-color-active: var(--mg-b-color-app);
  --mg-c-tabs-border-bottom: var(--mg-b-size-border) solid var(--mg-b-color-dark-soft);
}
