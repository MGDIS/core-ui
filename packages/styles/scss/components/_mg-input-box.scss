@use '../functions' as functions;

/**
 * Style for input "box":
 * - input[type="text"]
 * - input[type="date"]
 * - input[type="password"]
 * - textarea
 * - select
 */

.mg-c-input {
  &__box {
    height: var(--mg-b-size-min-height);
    box-sizing: border-box;
    padding: calc((var(--mg-b-size-min-height) - var(--mg-b-font-size) * var(--mg-b-line-height)) / 2) var(--mg-b-size-8);
    border-width: var(--mg-c-input-border-width);
    border-style: solid;
    border-color: var(--mg-c-input-color-border);
    border-radius: var(--mg-c-input-border-radius);
    background-color: var(--mg-b-color-light);
    color: var(--mg-b-color-dark);
    font-family: inherit;
    font-size: var(--mg-b-font-size);
    text-align: var(--mg-c-input-text-align, left);

    // placeholder
    &::placeholder {
      color: var(--mg-c-input-placeholder-color-text);
      font-style: italic;
    }

    &:focus {
      box-shadow: 0 0 var(--mg-b-size-4) var(--mg-c-input-color-box-shadow-focus);
    }

    // disabled
    &:disabled {
      opacity: var(--mg-b-opacity-element-disabled);
    }

    // Error
    .mg-c-input.mg-c-input--has-error & {
      border-color: var(--mg-b-color-danger);
    }
  }

  /**
  * Variations
  */
  &.mg-c-input--is-input-group-append {
    // Input group button
    --mg-c-button-border-radius-top-left: 0;
    --mg-c-button-border-radius-bottom-left: 0;

    /**
    * Append input slot
    */
    .mg-c-input__box {
      margin-right: calc(-1 * var(--mg-b-size-border));
      border-bottom-right-radius: 0;
      border-top-right-radius: 0;

      &:focus-visible {
        outline-offset: -0.2rem;
        outline-style: solid;
        outline-width: 0.2rem;
      }
    }
  }

  &.mg-c-input--has-buttons-group-append {
    ::slotted([slot='append-input']) {
      --mg-c-button-border-radius-top-right: 0;
      --mg-c-button-border-radius-bottom-right: 0;
    }

    ::slotted([slot='append-input']:not(:last-of-type)) {
      --mg-c-button-border-right-width: 0;
    }

    ::slotted([slot='append-input']:last-of-type) {
      --mg-c-button-border-radius-top-right: var(--mg-c-input-border-radius);
      --mg-c-button-border-radius-bottom-right: var(--mg-c-input-border-radius);
    }
  }

  &.mg-c-input--is-append-input-slot-content:not(.mg-c-input--readonly) ::slotted([slot='append-input']) {
    align-self: start;
    padding-top: calc((var(--mg-b-size-min-height) - functions.get-icon-size('medium')) / 2);
  }
}
