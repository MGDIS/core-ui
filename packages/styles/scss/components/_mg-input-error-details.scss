.mg-c-input__error {
  --mg-c-details-spacing: 0;

  &-details {
    margin: 0; // unset <p> default style
  }
}
