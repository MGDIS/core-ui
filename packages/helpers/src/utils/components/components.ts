import { isObject } from '../object';
import { CursorType, IGetPage } from './components.conf';

/**
 * Use to process code next tick in the event loop
 * @param callback - code to excute on next tick
 * @returns differed code excution
 */
export const nextTick = async (callback?: () => void): Promise<void> => {
  if (callback) return callback();
};

/**
 * Cursor possible values
 */
export const Cursor: Record<string, CursorType> = {
  FIRST: 'first',
  NEXT: 'next',
  PREVIOUS: 'previous',
  LAST: 'last',
} as const;

const DEFAULT_TOP = 10;

/**
 * Define a valid Page object and navigate throw page items with cursor.
 * Page object entries follow the REST API page practices.
 */
export class Page<T> {
  /**
   * Define items
   */
  public items: T[] = [];
  /**
   * Define total
   */
  public total: number;
  /**
   * Define top
   */
  public top: number = DEFAULT_TOP;
  /**
   * Define next
   */
  public next?: IGetPage<T> | string | URL;
  /**
   * Define base index
   */
  public readonly baseIndex = 1;

  constructor(init: Partial<Pick<Page<T>, 'items' | 'top' | 'total' | 'next'>>) {
    if (!isObject(init)) {
      throw new Error('Page - init must match IPage type.');
    } else {
      if (Array.isArray(init.items)) this.items = init.items;
      if (typeof init.top === 'number') this.top = init.top;
      this.total = typeof init.total === 'number' ? init.total : this.items.length;
      this.next = init.next;
    }
  }

  /**
   * Get index of items from cursor
   * @param cursor - cursor to find
   * @param oldItem - previous item
   * @returns item index
   */
  public getIndexFromCursor = (cursor: CursorType = 'first', oldItem?: T): number | null => {
    const startIndex = 0;
    if (!Array.isArray(this.items) || !this.items.length) return null;
    const lastIndex = this.items.length - this.baseIndex;

    let newIndex;
    let oldIndex = startIndex;
    if (['previous', 'next'].includes(cursor) && oldItem) {
      const findedIndex = this.items.findIndex(item => JSON.stringify(item) === JSON.stringify(oldItem));
      if (findedIndex === -1) return startIndex;
      oldIndex = findedIndex;
    }
    // Update index from cursor
    if (cursor === 'first') {
      newIndex = startIndex;
    } else if (cursor === 'last') {
      newIndex = lastIndex;
    } else if (cursor === 'previous') {
      newIndex = JSON.stringify(this.items[oldIndex]) === JSON.stringify(this.items[startIndex]) ? lastIndex : oldIndex - this.baseIndex;
    } else if (cursor === 'next') {
      newIndex = JSON.stringify(this.items[oldIndex]) === JSON.stringify(this.items[lastIndex]) ? startIndex : oldIndex + this.baseIndex;
    } else {
      newIndex = startIndex;
    }

    return newIndex;
  };
}

/**
 * Paginate an items array to navigate into with pages.
 * It follow REST standard and allow to navigate in items array with a similar format.
 */
export class Paginate<T> {
  /**
   * Define paginated items
   */
  public items: Page<T>['items'] = [];

  /* Privates */
  #top: Page<T>['top'] = DEFAULT_TOP;
  #next?: Page<T>['next'];
  #total?: Page<T>['total'];

  constructor(items: Page<T>['items'], options?: { step?: number; top?: Page<T>['top']; total?: Page<T>['total']; next?: Page<T>['next'] }) {
    if (Array.isArray(items)) this.items = items;
    if (options && (['string', 'function'].includes(typeof options.next) || (isObject<URL>(options.next) && URL.canParse(options.next)))) this.#next = options.next;
    if (typeof options?.top === 'number') this.#top = options.top;
    if (typeof options?.total === 'number') this.#total = options.total;
  }

  /**
   * Get page
   * @param offset - pagiantion offset
   * @param filter - filter methode
   * @returns formated page
   */
  public getPage: IGetPage<T> = (offset = 0, filter?) => {
    const items = typeof filter === 'function' ? this.items.filter(filter) : this.items;
    let next;
    if (this.#next) next = this.#next;
    else if (items.length > offset + this.#top) next = () => this.getPage(offset + this.#top, filter);

    return new Page({
      items: items.slice(offset, offset + this.#top),
      total: this.#total,
      top: this.#top,
      next,
    });
  };
}
