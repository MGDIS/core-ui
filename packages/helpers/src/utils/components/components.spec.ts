import { describe, expect, test, vi } from 'vitest';
import { Cursor, nextTick, Page, Paginate } from './components';
import { CursorType } from './components.conf';

const initArray = length => new Array(length).fill('').map((_, key) => (key + 1).toString());

describe('components', () => {
  describe('nextTick', () => {
    test('Should get a promise', async () => {
      const job = nextTick();
      expect(job.then).toBeDefined();

      const res = await job;
      expect(res).toBeUndefined();
    });
    test('Should wrapp and execute a callback in a promise', async () => {
      const fn = vi.fn();
      const job = nextTick(fn);
      expect(job.then).toBeDefined();

      const res = await job;
      expect(fn).toHaveBeenCalled();
      expect(res).toBeUndefined();
    });
  });
  describe('Paginate', () => {
    test('Should create Paginate instance', () => {
      const res = new Paginate([]);
      expect(res.items.length).toEqual(0);
      expect(res).toHaveProperty('getPage');
    });

    describe('getPage()', () => {
      test.each([
        {
          items: {},
          options: undefined,
        },
        {
          items: [],
          options: undefined,
        },
        {
          items: ['batman', 'robin', 'joker', 'bane', 'harley quinn', 'dante'],
          options: undefined,
        },
        {
          items: initArray(34),
          options: undefined,
        },
        {
          items: initArray(34),
          options: { top: 5 },
        },
        {
          items: initArray(34),
          options: { total: 105 },
        },
        {
          items: initArray(34),
          options: { next: '/next/page' },
        },
        {
          items: initArray(34),
          options: { next: new URL('http://url.fr/next/page') },
        },
      ])('Shoud get paginated page from instance %s', ({ items, options }) => {
        const res = new Paginate(items as unknown[], options);
        const page = res.getPage();
        expect(page).toMatchSnapshot();
      });

      test('Should filter paginated page from instance', () => {
        const res = new Paginate(['batman', 'robin', 'joker', 'bane', 'harley quinn', 'dante']);
        const page = res.getPage(0, val => val.includes('an'));
        expect(page).toMatchSnapshot();
      });

      test('Should get next() page from instance', () => {
        const res = new Paginate(initArray(20));
        // get first page
        const page = res.getPage();
        expect(page).toMatchSnapshot();
        // get next page
        const next = (page.next as unknown as () => Page<unknown>)();
        expect(next).toMatchSnapshot();
      });
    });
  });

  describe('Page', () => {
    const items = ['batman', 'robin', 'joker', 'bane', 'harley quinn', 'dante'];
    test.each([
      {
        items,
      },
      {
        items: items.map((item, index) => ({ title: item, value: index })) as unknown[],
      },
      {
        items,
        total: 100,
      },
      {
        items,
        top: 10,
      },
      {
        items,
        next: '/next/url',
      },
    ])('Should create Page instance', init => {
      const page = new Page(init);
      expect(page).toMatchSnapshot();
    });

    test('Should thrown error with invalid Page init', () => {
      expect.assertions(1);
      try {
        new Page('batman' as unknown as Record<string, string>);
      } catch (err) {
        expect(err.message).toEqual('Page - init must match IPage type.');
      }
    });

    describe('getIndexFromCursor()', () => {
      test.each([undefined, 'hello'])('Should get first index by default, cursor %s', cursor => {
        const page = new Page({ items });
        page.items.forEach(() => {
          const index = page.getIndexFromCursor(cursor as CursorType);
          expect(index).toEqual(0);
        });
      });
      describe.each(Object.values(Cursor))('cursor %s', cursor => {
        let result = 0;
        if (cursor === 'first') result = 0;
        if (cursor === 'last') result = items.length - 1;
        if (cursor === 'next') result = 1;
        if (cursor === 'previous') result = items.length - 1;
        test.each([
          {
            items: 'batman',
            result: null,
          },
          {
            items: {},
            result: null,
          },
          {
            items: [],
            result: null,
          },
          {
            items,
            result,
          },
          {
            items: items.map((item, index) => ({ title: item, value: index })) as unknown[],
            result,
          },
        ])('Sould get index from curor %s', ({ items, result }) => {
          const page = new Page({ items: items as unknown[] });
          const index = page.getIndexFromCursor(cursor);
          expect(index).toEqual(result);
        });
      });
      test('Should go to next value witdh oldValue', () => {
        const page = new Page({ items });
        for (const oldIndex in page.items) {
          const index = page.getIndexFromCursor('next', items[Number(oldIndex)]);
          const nextIndex = Number(oldIndex) + page.baseIndex;
          expect(index).toEqual(nextIndex >= page.items.length ? 0 : nextIndex);
        }
      });
      test('Should go to previous value witdh oldValue', () => {
        const page = new Page({ items });
        for (const oldIndex in page.items) {
          const index = page.getIndexFromCursor('previous', items[Number(oldIndex)]);
          const nextIndex = Number(oldIndex) - page.baseIndex;
          expect(index).toEqual(Number(oldIndex) === 0 ? page.items.length - page.baseIndex : nextIndex);
        }
      });
      test.each(['previous', 'next'])('Should get first index with unavailable item, cursor %s', cursor => {
        const page = new Page({ items });
        page.items.forEach(() => {
          const index = page.getIndexFromCursor(cursor as CursorType, 'hello');
          expect(index).toEqual(0);
        });
      });
    });
  });
});
