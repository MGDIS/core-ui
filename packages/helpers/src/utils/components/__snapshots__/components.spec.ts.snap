// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components > Page > Should create Page instance 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "batman",
    "robin",
    "joker",
    "bane",
    "harley quinn",
    "dante",
  ],
  "next": undefined,
  "top": 10,
  "total": 6,
}
`;

exports[`components > Page > Should create Page instance 2`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    {
      "title": "batman",
      "value": 0,
    },
    {
      "title": "robin",
      "value": 1,
    },
    {
      "title": "joker",
      "value": 2,
    },
    {
      "title": "bane",
      "value": 3,
    },
    {
      "title": "harley quinn",
      "value": 4,
    },
    {
      "title": "dante",
      "value": 5,
    },
  ],
  "next": undefined,
  "top": 10,
  "total": 6,
}
`;

exports[`components > Page > Should create Page instance 3`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "batman",
    "robin",
    "joker",
    "bane",
    "harley quinn",
    "dante",
  ],
  "next": undefined,
  "top": 10,
  "total": 100,
}
`;

exports[`components > Page > Should create Page instance 4`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "batman",
    "robin",
    "joker",
    "bane",
    "harley quinn",
    "dante",
  ],
  "next": undefined,
  "top": 10,
  "total": 6,
}
`;

exports[`components > Page > Should create Page instance 5`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "batman",
    "robin",
    "joker",
    "bane",
    "harley quinn",
    "dante",
  ],
  "next": "/next/url",
  "top": 10,
  "total": 6,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [ '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34' ], options: { next: '/next/page' } } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "next": "/next/page",
  "top": 10,
  "total": 10,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [ '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34' ], options: { next: URL{ [Symbol(impl)]: URLImpl{ _url: { scheme: 'http', username: '', password: '', host: 'url.fr', port: null, path: [ 'next', 'page' ], query: null, fragment: null }, _query: URLSearchParamsImpl{ _list: [], _url: URLImpl[Circular], [Symbol(wrapper)]: URLSearchParams{ [Symbol(impl)]: URLSearchParamsImpl[Circular] } }, [Symbol(wrapper)]: URL[Circular] } } } } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "next": "http://url.fr/next/page",
  "top": 10,
  "total": 10,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [ '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34' ], options: { top: 5 } } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "1",
    "2",
    "3",
    "4",
    "5",
  ],
  "next": [Function],
  "top": 5,
  "total": 5,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [ '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34' ], options: { total: 105 } } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "next": [Function],
  "top": 10,
  "total": 105,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [ '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34' ], options: undefined } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "next": [Function],
  "top": 10,
  "total": 10,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [ 'batman', 'robin', 'joker', 'bane', 'harley quinn', 'dante' ], options: undefined } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "batman",
    "robin",
    "joker",
    "bane",
    "harley quinn",
    "dante",
  ],
  "next": undefined,
  "top": 10,
  "total": 6,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: [], options: undefined } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [],
  "next": undefined,
  "top": 10,
  "total": 0,
}
`;

exports[`components > Paginate > getPage() > Shoud get paginated page from instance { items: {}, options: undefined } 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [],
  "next": undefined,
  "top": 10,
  "total": 0,
}
`;

exports[`components > Paginate > getPage() > Should filter paginated page from instance 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "batman",
    "bane",
    "dante",
  ],
  "next": undefined,
  "top": 10,
  "total": 3,
}
`;

exports[`components > Paginate > getPage() > Should get next() page from instance 1`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "10",
  ],
  "next": [Function],
  "top": 10,
  "total": 10,
}
`;

exports[`components > Paginate > getPage() > Should get next() page from instance 2`] = `
Page {
  "baseIndex": 1,
  "getIndexFromCursor": [Function],
  "items": [
    "11",
    "12",
    "13",
    "14",
    "15",
    "16",
    "17",
    "18",
    "19",
    "20",
  ],
  "next": undefined,
  "top": 10,
  "total": 10,
}
`;
